netket/operator/abstract_operator.py

```diff
@@ -0,0 +1,125 @@
+                array: An array containing the matrix elements :math:`O(x,x')` associated to each x'.
+
+        """
+        return NotImplementedError
+
+    def n_conn(self, x, out=None):
+        r"""Return the number of states connected to x.
+
+            Args:
+                x (matrix): A matrix of shape (batch_size,hilbert.size) containing
+                            the batch of quantum numbers x.
+                out (array): If None an output array is allocated.
+
+            Returns:
+                array: The number of connected states x' for each x[i].
+
+        """
+        if out is None:
+            out = _np.empty(x.shape[0], dtype=_np.intc)
+        self.get_conn_flattened(x, out)
+        out = self._n_conn_from_sections(out)
+
+        return out
+
+    @staticmethod
+    @jit(nopython=True)
+    def _n_conn_from_sections(out):
+        low = 0
+        for i in range(out.shape[0]):
+            old_out = out[i]
+            out[i] = out[i] - low
+            low = old_out
+
+        return out
+
+    @property
+    @abc.abstractmethod
+    def hilbert(self):
+        r"""AbstractHilbert: The hilbert space associated to this operator."""
+        return NotImplementedError
+
+    def to_sparse(self):
+        r"""Returns the sparse matrix representation of the operator. Note that,
+            in general, the size of the matrix is exponential in the number of quantum
+            numbers, and this operation should thus only be performed for
+            low-dimensional Hilbert spaces or sufficiently sparse operators.
+
+            This method requires an indexable Hilbert space.
+
+            Returns:
+                scipy.sparse.csr_matrix: The sparse matrix representation of the operator.
+        """
+        hilb = self.hilbert
+
+        x = hilb.all_states()
+
+        sections = _np.empty(x.shape[0], dtype=_np.int32)
+        x_prime, mels = self.get_conn_flattened(x, sections)
+
+        numbers = hilb.states_to_numbers(x_prime)
+
+        sections1 = _np.empty(sections.size + 1, dtype=_np.int32)
+        sections1[1:] = sections
+        sections1[0] = 0
+
+        return _csr_matrix((mels, numbers, sections1))
+
+    def to_dense(self):
+        r"""Returns the dense matrix representation of the operator. Note that,
+            in general, the size of the matrix is exponential in the number of quantum
+            numbers, and this operation should thus only be performed for
+            low-dimensional Hilbert spaces or sufficiently sparse operators.
+
+            This method requires an indexable Hilbert space.
+
+            Returns:
+                numpy.matrix: The dense matrix representation of the operator.
+        """
+        return self.to_sparse().todense()
+
+    def apply(self, v):
+        return self.to_sparse().dot(v)
+
+    def __call__(self, v):
+        return self.apply(v)
+
+    def to_linear_operator(self):
+        return self.to_sparse()
```

Comment:
This should probably be implemented by computing matrix elements on-the-fly before merging this PR. All that's needs to be minimally compatible with [scipy.sparse.linalg.LinearOperator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html) would be a method `matvec` implementing the matrix-vector product, then we'd just return `LinearOperator(shape, matvec)`.

Reply:
yes we could implement this as done in the previous versions. In practice however sparse matrices are much faster when doing matrix vector multiplications, and (typically) only have a constant memory penalty over the linear operator, so I am not sure it is really worth adding this?